drivers/watchdog/wdt_opentitan.c

```diff
@@ -0,0 +1,178 @@
+		/* Force a read to ensure the timer was enabled. */
+		(void) regs->wdog_ctrl;
+		regs->wdog_regwen = 0;
+	}
+	return 0;
+}
+
+static int ot_aontimer_disable(const struct device *dev)
+{
+	volatile struct ot_aontimer_regs *regs = REG_OT_WDT(dev);
+
+	if (!regs->wdog_regwen) {
+		LOG_ERR("Cannot disable - watchdog settings locked.");
+		return -EPERM;
+	}
+
+	if (!(regs->wdog_ctrl & BIT(0))) {
+		return -EFAULT;
+	}
+	regs->wdog_ctrl &= ~BIT(0);
+
+	return 0;
+}
+
+/*
+ * The OpenTitan AON Timer includes a multi-level watchdog timer.
+ * While the first stage supports an interrupt callback, the second does not.
+ * The second stage is mandatory to adjust the "bite" time window.
+ *
+ * Some boundaries are enforced to prevent behavior that is technically correct
+ * but is not likely intended.
+ * Bark:
+ * Minimum must be 0. Maximum must be > min.
+ * The bark interrupt occurs at max.
+ * Bite:
+ * Minimum must be >= bark.minimum. Maximum > min.
+ * The bite action occurs max.
+ */
+static int ot_aontimer_install_timeout(const struct device *dev,
+					const struct wdt_timeout_cfg *cfg)
+{
+	const struct wdt_ot_aontimer_cfg *const wdt_cfg = dev->config;
+	struct wdt_ot_aontimer_data *data = dev->data;
+	volatile struct ot_aontimer_regs *regs = REG_OT_WDT(dev);
+	struct wdt_timeout_cfg *bite = cfg->next;
+	uint64_t bark_thold, bite_thold;
+
+	if (!regs->wdog_regwen) {
+		LOG_ERR("Cannot install timeout - watchdog settings locked.");
+		return -ENOMEM;
+	}
+
+	if (bite == NULL || bite->window.max < cfg->window.max) {
+		return -EINVAL;
+	}
+
+	/* Force 64-bit ops to ensure thresholds fits in the timer reg. */
+	bite_thold = ((uint64_t) bite->window.max * wdt_cfg->clk_freq / 1000);
+	bark_thold = (cfg->window.max * wdt_cfg->clk_freq / 1000);
+	if (bite_thold > UINT32_MAX || bark_thold > UINT32_MAX) {
+		return -EINVAL;
+	}
+
+	data->bark = cfg->callback;
+
+	regs->wdog_bark_thold = (uint32_t) bark_thold;
+	regs->wdog_bite_thold = (uint32_t) bite_thold;
+
+	return 0;
+}
+
+static int ot_aontimer_feed(const struct device *dev, int channel_id)
+{
+	ARG_UNUSED(channel_id);
+	volatile struct ot_aontimer_regs *regs = REG_OT_WDT(dev);
+
+	regs->wdog_count = 0;
+
+	/* Deassert the interrupt line */
+	regs->intr_state = BIT(1);
+	return 0;
+}
+
+static void wdt_ot_isr(const struct device *dev)
+{
+	struct wdt_ot_aontimer_data *data = dev->data;
+	volatile struct ot_aontimer_regs *regs = REG_OT_WDT(dev);
+
+	if (data->bark != NULL) {
+		data->bark(dev, 0);
+	}
+
+	/* Deassert the interrupt line */
+	regs->intr_state = BIT(1);
+}
+
+static int ot_aontimer_init(const struct device *dev)
+{
+	IRQ_CONNECT(DT_IRQ_BY_IDX(NODE_AON, 0, irq),
+		DT_IRQ_BY_IDX(NODE_AON, 0, priority), wdt_ot_isr,
+		DEVICE_DT_INST_GET(0), 0);
+	irq_enable(DT_IRQ_BY_IDX(NODE_AON, 0, irq));
+
+	return 0;
+}
+
+static struct wdt_ot_aontimer_data ot_aontimer_data;
+
+static struct wdt_ot_aontimer_cfg ot_aontimer_cfg = {
+	.regs = (struct ot_aontimer_regs *)DT_INST_REG_ADDR(0),
```

Comment:
you will have always one and only one instance of this?

Reply:
I believe so, yes. I don't think a board will have more than one AON Timer instantiation.

Reply:
right, but why do not simply use `DT_INST_FOREACH_STATUS_OKAY` as usual? That would be more generic than hardcoding one singleton.

Reply:
I modeled the driver after other watchdog drivers, which seem to hardcode the single instance. Generalizing it would make the implementation more complicated because of the interrupt feature.

Reply:
@gmarull @mbolivar do we have any strong indication about this?

Reply:
It's ok to have singletons, as soon as driver code doesn't use DT macros directly inside.